bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2ThingHandler.java

```diff
@@ -631,12 +631,38 @@ public void initialize() {
         }
     }
 
+    /**
+     * Update the channel state depending on new resources sent from the bridge.
+     *
+     * @param resources a collection of Resource objects containing the new state.
+     */
+    public void onResources(Collection<Resource> resources) {
```

Comment:
Ah yes. That is certainly a step forward. And it is not so 'massive' as I had feared in the darkness of last night. I still have a few comments though..

- On line 641 and 642 you are missing a pair of brackets `ww && xx && (yy || zz)` 
- On line 643 using a forEach lambda might improve the performance a bit. ??
- On line 645 your code is still wrong; you only want to skip channel updates for deactivation events, yet your current code skips channel updates for both activation and deactivation events.
 
I said earlier that I had some thoughts about adding a Junit test. I was thinking of using static methods with an interface in the Clip2ThingHandler so it could call back a mocked inner onResource() method when running in a Junit context. But this is after all not so easy either. So there is still no Junit test that would allow you verify my assertion concerning line 645 above. As you know, I wrote a rough test suite that uses System.out.println() statements in order to verify `onResources()` and I have modified this to test both your and my current versions of onResources(). And you could choose to accept it from me that my version does pass those tests, whereas yours does not.

```
    public void onResources(Collection<Resource> resources) {
        boolean sceneActivated = resources.stream().anyMatch(r -> sceneContributorsCache.containsKey(r.getId())
                && (r.getSceneActive().orElse(false) || r.getSmartSceneActive().orElse(false)));
        resources.forEach(r -> {
            boolean skipUpdate = sceneActivated && sceneContributorsCache.containsKey(r.getId())
                    && (!r.getSceneActive().orElse(true) || !r.getSmartSceneActive().orElse(true));
            onResource(r, !skipUpdate);
        });
    }
```

Reply:
> Ah yes. That is certainly a step forward. And it is not so 'massive' as I had feared in the darkness of last night. I still have a few comments though..

Yeah, I figured we just missed one condition, i.e. to check if the resource having scene fields belongs to the Thing. And the cache can answer this question.

> * On line 641 and 642 you are missing a pair of brackets `ww && xx && (yy || zz)`

It was unfinished, so I did one more push. 🙂 Can you check again? I tested it manually and the current version works.

> * On line 643 using a forEach lambda might improve the performance a bit. ??

Actually it's the other way around.

> * On line 645 your code is still wrong; you only want to skip channel updates for deactivation events, yet your current code skips channel updates for both activation and deactivation events.

Yes, I also realized that right after pushing, and marked as draft. This was a natural consequence of setting `sceneActivated` before the loop, so it had to be refactored. It should be okay in latest version.

> I said earlier that I had some thoughts about adding a Junit test. I was thinking of using static methods with an interface in the Clip2ThingHandler so it could call back a mocked inner onResource() method when running in a Junit context. But this is after all not so easy either. So there is still no Junit test that would allow you verify my assertion concerning line 645 above. As you know, I wrote a rough test suite that uses System.out.println() statements in order to verify `onResources()` and I have modified this to test both your and my current versions of onResources(). And you could choose to accept it from me that my version does pass those tests, whereas yours does not.

I will have a look later today. If this means the tests would have to be run and checked manually by a developer rather than by CI, I'm not sure how much value it provides? Since there is no test coverage of the thing handler now anyway, we could go ahead without for 4.1, and add proper test coverage later. There are many bindings to take inspiration from with regards to proper mocking and testing of handlers.

Reply:
>> It was unfinished, so I did one more push. 🙂 Can you check again?

Unfortunately it is still not right. I know that you have a thing about removing excess brackets in complex logic operations. And frankly I am not 100% sure about the precedence order of `||` vs `&&`. But in this case I think it really helps to use more brackets than fewer in order to make the logic explicit rather than implicit..

```
// yours
.anyMatch(r -> sceneContributorsCache.containsKey(r.getId()) && r.getSceneActive().orElse(false) || r.getSmartSceneActive().orElse(false));
// mine
.anyMatch(r -> sceneContributorsCache.containsKey(r.getId()) && (r.getSceneActive().orElse(false) || r.getSmartSceneActive().orElse(false)));
```

``` 
// yours
boolean skipUpdate = sceneActivated
                    && !(resource.getSceneActive().orElse(false) || resource.getSmartSceneActive().orElse(false));
// mine
boolean skipUpdate = sceneActivated && sceneContributorsCache.containsKey(r.getId())
                    && (!r.getSceneActive().orElse(true) || !r.getSmartSceneActive().orElse(true));
```

Note: concerning the latter above: I am not sure if your outer inversion with orElse(false) does actually evaluate identical to my inner inversion with orElse(true) .. maybe / maybe not. But I can tell you that my version has at least been tested manually.

Reply:
That was a clear bug, thanks. Fixed.

Reply:
Apropos Junit testing..
1. I have no objection to postponing the introduction of a proper Junit test
2. Nevertheless once you are ready I will still do a final manual test before I approve the PR

Reply:
> once you are ready I will still do a final manual test before I approve the PR

@jlaur let me know when you are ready (the current version fails my tests)

Reply:
How does it fail?

Reply:
> How does it fail?

See below.
The problem is the logic in your line 644 .. where I have already posted my suggestion above.

```
testTransitionA
jlaur code
Resource id:0, type:light => skipped
Resource id:1, type:scene => processed
Resource id:2, type:light => skipped
Resource id:3, type:scene => skipped
Resource id:4, type:light => skipped
andrewfg code
Resource id:0, type:light => processed
Resource id:1, type:scene => processed
Resource id:2, type:light => processed
Resource id:3, type:scene => skipped
Resource id:4, type:light => processed

testTransitionB
jlaur code
Resource id:0, type:light => skipped
Resource id:1, type:scene => skipped
Resource id:2, type:light => skipped
Resource id:3, type:scene => processed
Resource id:4, type:light => skipped
andrewfg code
Resource id:0, type:light => processed
Resource id:1, type:scene => skipped
Resource id:2, type:light => processed
Resource id:3, type:scene => processed
Resource id:4, type:light => processed

testTransitionC
jlaur code
Resource id:0, type:light => skipped
Resource id:1, type:scene => processed
Resource id:2, type:light => skipped
Resource id:3, type:smart_scene => skipped
Resource id:4, type:light => skipped
andrewfg code
Resource id:0, type:light => processed
Resource id:1, type:scene => processed
Resource id:2, type:light => processed
Resource id:3, type:smart_scene => skipped
Resource id:4, type:light => processed

testTransitionD
jlaur code
Resource id:0, type:light => skipped
Resource id:1, type:smart_scene => processed
Resource id:2, type:light => skipped
Resource id:3, type:scene => skipped
Resource id:4, type:light => skipped
andrewfg code
Resource id:0, type:light => processed
Resource id:1, type:smart_scene => processed
Resource id:2, type:light => processed
Resource id:3, type:scene => skipped
Resource id:4, type:light => processed

testTransitionE
jlaur code
Resource id:0, type:light => skipped
Resource id:1, type:smart_scene => processed
Resource id:2, type:light => skipped
Resource id:3, type:smart_scene => skipped
Resource id:4, type:light => skipped
andrewfg code
Resource id:0, type:light => processed
Resource id:1, type:smart_scene => processed
Resource id:2, type:light => processed
Resource id:3, type:smart_scene => skipped
Resource id:4, type:light => processed

testTransitionF
jlaur code
Resource id:0, type:light => skipped
Resource id:1, type:scene => processed
Resource id:2, type:light => skipped
Resource id:3, type:scene => processed
Resource id:4, type:light => skipped
andrewfg code
Resource id:0, type:light => processed
Resource id:1, type:scene => processed
Resource id:2, type:light => processed
Resource id:3, type:scene => processed
Resource id:4, type:light => processed

testTransitionG
jlaur code
Resource id:0, type:light => skipped
Resource id:1, type:smart_scene => processed
Resource id:2, type:light => skipped
Resource id:3, type:smart_scene => processed
Resource id:4, type:light => skipped
andrewfg code
Resource id:0, type:light => processed
Resource id:1, type:smart_scene => processed
Resource id:2, type:light => processed
Resource id:3, type:smart_scene => processed
Resource id:4, type:light => processed

testTransitionH
jlaur code
Resource id:0, type:light => processed
Resource id:1x, type:scene => processed
Resource id:2, type:light => processed
Resource id:3, type:scene => processed
Resource id:4, type:light => processed
andrewfg code
Resource id:0, type:light => processed
Resource id:1x, type:scene => processed
Resource id:2, type:light => processed
Resource id:3, type:scene => processed
Resource id:4, type:light => processed

testTransitionI
jlaur code
Resource id:0, type:light => skipped
Resource id:1, type:scene => processed
Resource id:2, type:light => skipped
Resource id:3x, type:scene => skipped
Resource id:4, type:light => skipped
andrewfg code
Resource id:0, type:light => processed
Resource id:1, type:scene => processed
Resource id:2, type:light => processed
Resource id:3x, type:scene => processed
Resource id:4, type:light => processed
```

Reply:
Sorry for being slow, you are right. I have given it another go inverting some things to (hopefully) make it simpler/easier to read without double negations. Does it parse your tests, and how do you like this version?

Reply:
@jlaur thanks for the update; I will run the tests again and give (hopefully) final approval tomorrow.

Reply:
@jlaur I checked your code. I think the logic inversion is a good improvement, and the left to right 'or' will probably improve performance. However I think there is still an issue concerning the two 'orElse(false)' statements, which with the logic inversion should be changed to 'orElse(true)'. This handles the cases where there may be an empty scene resource that contains neither an activate nor a deactivate element (but perhaps something else). I think those cases should be allowed through the gate to update the channels too. The updateChannels method currently has no use for such resources, but I think we should not close the gate via this code here, but rather let updateChannels make that decision.

Reply:
^
HOWEVER I realise that simply changing simply to 'orElse(true)' is also wrong, since the two combined would always cause the resource to pass through to updateChannels. Aargh!

Reply:
I think if there would be "something else" related to scenes (other than currently active scene) that we would need to handle, this logic would need to be adapted/rewritten anyway, since it's based on the assumption that we are only updating that channel. So I'm wondering if there is actually anything reasonable to do at this moment?

Reply:
Yeah. The 'something else' case would be when it is a scene resource and both getSceneActive and getSmartSceneActive are empty.

Reply:
So let us leave the code as it is for the moment.